File: frontend/src/types/investment.ts
export interface DemographicProfile {
  locality: string;
  postcodes: string[];
  population_2016: number;
  population_2021: number;
  population_growth_5y_pct: number;
  population_growth_annualised_pct: number;
  median_household_income_2016: number;
  median_household_income_2021: number;
  income_growth_5y_pct: number;
  income_growth_annualised_pct: number;
  latitude: number;
  longitude: number;
  data_as_of: number;
  geography_level: string;
}

export interface ScoreComponent {
  key: string;
  label: string;
  raw_value: number;
  percentile_score: number;
  weight_pct: number;
  contribution: number;
  explanation: string;
}

export interface InvestmentScore {
  locality: string;
  postcodes: string[];
  latitude: number;
  longitude: number;
  median_weekly_rent: number;
  rent_growth_1y_pct: number;
  rental_count: number;
  median_sale_price: number;
  price_growth_5y_annualised_pct: number;
  sales_count: number;
  gross_yield_pct: number;
  population_2021: number;
  population_growth_5y_pct: number;
  median_household_income_weekly: number;
  income_growth_5y_pct: number;
  investment_score: number;
  confidence: string;
  rental_data_as_of: string;
  sales_data_as_of: string;
  demographic_data_as_of: number;
  components: ScoreComponent[];
}

export type RankingSort =
  | "score"
  | "yield"
  | "price_growth"
  | "rent_growth"
  | "population_growth"
  | "income_growth";

export interface RankingFilters {
  maxPrice: number;
  minYield: number;
  minScore: number;
  search?: string;
  sort: RankingSort;
  limit?: number;
}

File: backend/models/investment.py
"""Demographic and explainable investment-ranking models."""

from dataclasses import dataclass
from datetime import date


@dataclass(frozen=True)
class DemographicProfile:
    locality: str
    postcodes: list[str]
    population_2016: int
    population_2021: int
    population_growth_5y_pct: float
    population_growth_annualised_pct: float
    median_household_income_2016: float
    median_household_income_2021: float
    income_growth_5y_pct: float
    income_growth_annualised_pct: float
    latitude: float
    longitude: float
    data_as_of: int = 2021
    geography_level: str = "ABS Postal Area"


@dataclass(frozen=True)
class ScoreComponent:
    key: str
    label: str
    raw_value: float
    percentile_score: float
    weight_pct: float
    contribution: float
    explanation: str


@dataclass(frozen=True)
class InvestmentScore:
    locality: str
    postcodes: list[str]
    latitude: float
    longitude: float
    median_weekly_rent: float
    rent_growth_1y_pct: float
    rental_count: int
    median_sale_price: float
    price_growth_5y_annualised_pct: float
    sales_count: int
    gross_yield_pct: float
    population_2021: int
    population_growth_5y_pct: float
    median_household_income_weekly: float
    income_growth_5y_pct: float
    investment_score: float
    confidence: str
    rental_data_as_of: date
    sales_data_as_of: date
    demographic_data_as_of: int
    components: list[ScoreComponent]
